Added `run_on_start` feature to scheduler

DenuxPlays opened the pull request:
Fixes: #1182

DenuxPlays commented:
Going to fix CI.
Also @kaplanelad is there a reason that in the run function `tokio::signal::ctrl_c().await?;` is used instead of `sched.shutdown_on_ctrl_c();`?

kaplanelad reviewed (changes requested):
We can simplify the implementation as follows:

First, we can remove the `run_on_start` function. When starting the scheduler, the `run` function already handles all scheduled jobs, making `run_on_start` is redundant.

The `tokio-cron-scheduler` crate that we are using provides a [new_one_shot](https://github.com/mvniekerk/tokio-cron-scheduler/blob/main/src/job/mod.rs#L397) function. This allows a job to be triggered immediately and ensures it runs only once, while also supporting synchronized execution.

Expected implementation:
1. Call `run`, which manages all scheduler jobs (including async and one-time jobs). This eliminates the need for the `run_on_start` function.
2. In the job logic, check if the current job exists in the run_on_start configuration.
3. If it does, use `new_one_shot` to trigger the job immediately and ensure it only runs once.
4. Retain the `new_async` functionality for asynchronous job handling.



This implementation reduces code duplication, centralizes all job-triggering logic, and ensures the job executor handles all jobs efficiently and easy to test ().

kaplanelad commented:
> Going to fix CI. Also @kaplanelad is there a reason that in the run function `tokio::signal::ctrl_c().await?;` is used instead of `sched.shutdown_on_ctrl_c();`?

yes

DenuxPlays commented:
> We can simplify the implementation as follows:
> 
> First, we can remove the `run_on_start` function. When starting the scheduler, the `run` function already handles all scheduled jobs, making `run_on_start` is redundant.
> 
> The `tokio-cron-scheduler` crate that we are using provides a [new_one_shot](https://github.com/mvniekerk/tokio-cron-scheduler/blob/main/src/job/mod.rs#L397) function. This allows a job to be triggered immediately and ensures it runs only once, while also supporting synchronized execution.
> 
> Expected implementation:
> 
> 1. Call `run`, which manages all scheduler jobs (including async and one-time jobs). This eliminates the need for the `run_on_start` function.
> 2. In the job logic, check if the current job exists in the run_on_start configuration.
> 3. If it does, use `new_one_shot` to trigger the job immediately and ensure it only runs once.
> 4. Retain the `new_async` functionality for asynchronous job handling.
> 
> This implementation reduces code duplication, centralizes all job-triggering logic, and ensures the job executor handles all jobs efficiently and easy to test ().

Addressed everything.
Also moved the execute job functionality into its own function to reduce code duplication and to ensure that we use the same logic.

kaplanelad reviewed (changes requested):
Thanks for addressing the previous changes!  I’ve just noticed a few more things 

**Documentation:**
Could you please add details about this feature to the [loco.rs documentation](https://loco.rs/docs/processing/scheduler/)?

**CLI:**
Running this command:
```sh
cargo loco scheduler --list
```
should display all scheduler configurations. Would it be possible to add a new column named `run on start` to the output? This column should indicate true or false based on the configuration.


**Validation**
What happens if a task ID that doesn’t exist is added to the run_on_start field? if do nothing, please add validation and make sure that the scheduler not starting to run

DenuxPlays commented:
For validation:

I would rather print a warning than stopping the scheduler.
This isn't a reason for the scheduler to not work (in most cases a typo or just forgotten to remove).
Would that be enough or do you really want the scheduler to stop with an error?

DenuxPlays commented:
I've moved the attribute to the job so you can take a look.
If this is unwanted I can revert this commit.

Docs are still missing

kaplanelad reviewed (approved):
Thanks, looks good now

DenuxPlays commented:
This can be merged then Right?
